In [1]:
import numpy as np

print(type(np))

<class 'module'>


The line import numpy as np is a common statement used in Python to import the NumPy library and assign it the alias np. Here’s a breakdown of what this line does:

Import Statement: import numpy: This instructs Python to load the NumPy library into your current environment. NumPy is a powerful numerical computing library in Python, widely used for array computations and mathematical operations.

Alias Assignment: as np: This part of the statement assigns an alias np to NumPy. This alias allows you to refer to NumPy functions and objects using np instead of typing out numpy each time, which can make your code more concise and readable.

Why Use NumPy?
Efficient Arrays: NumPy provides an array object that is more efficient than Python's built-in list for numerical operations.
Mathematical Functions: NumPy includes a wide range of mathematical functions for operations on these arrays.
Linear Algebra Operations: NumPy supports linear algebra operations, Fourier transforms, and random number generation.

NumPy (Numerical Python) is a fundamental library in Python for numerical computing. It provides support for large, multi-dimensional arrays and matrices, along with a collection of mathematical functions to operate on these arrays efficiently. Here are several reasons why NumPy is widely used and essential in the Python ecosystem:

### 1. **Efficient Array Operations:**
   - NumPy arrays are more efficient than Python lists for numerical operations. They are implemented in C and allow for vectorized operations, where operations are applied to whole arrays instead of individual elements. This reduces the need for explicit loops in Python, making code execution faster and more concise.

### 2. **Multi-dimensional Data Representation:**
   - NumPy arrays can represent multi-dimensional data (e.g., matrices or tensors) with ease. This makes NumPy suitable for handling datasets used in fields such as machine learning, scientific computing, and data analysis.

### 3. **Broad Range of Mathematical Functions:**
   - NumPy provides a rich collection of mathematical functions that operate on arrays. These include basic arithmetic operations, trigonometric functions, statistical functions, linear algebra routines, Fourier transforms, and more. These functions are optimized for performance and provide consistent results across different platforms.

### 4. **Integration with Other Libraries:**
   - NumPy is a foundational library in the Python scientific computing stack. It integrates well with other libraries like SciPy (Scientific Python) for advanced mathematical functions, Matplotlib for plotting, Pandas for data manipulation and analysis, and scikit-learn for machine learning tasks.

### 5. **Memory Efficiency:**
   - NumPy arrays are memory efficient compared to Python lists. They allow for better memory management and enable faster access and manipulation of data. This is crucial when working with large datasets that cannot fit into the computer's main memory.

### 6. **Support for Broadcasting:**
   - Broadcasting is a powerful feature in NumPy that allows operations to be performed on arrays of different shapes. NumPy automatically broadcasts these operations, making code more readable and reducing the need for explicit looping constructs.

### 7. **Platform for Scientific Computing:**
   - NumPy forms the basis for many scientific and numerical computing libraries in Python. Its consistent interface and robust functionality make it a preferred choice for researchers, engineers, and data scientists working on computational and data-intensive applications.

### Example Usage:
```python
import numpy as np

# Create a NumPy array
arr = np.array([1, 2, 3, 4, 5])

# Perform operations
mean_value = np.mean(arr)
print("Mean:", mean_value)
```

In this example:
- `np.array([1, 2, 3, 4, 5])` creates a NumPy array.
- `np.mean(arr)` calculates the mean of the array `arr` using a NumPy function.

### Conclusion:
NumPy's versatility, performance, and extensive mathematical functions make it indispensable for numerical computations and data manipulation in Python. Its efficient handling of arrays, integration with other libraries, and support for advanced operations ensure that it remains a cornerstone of scientific computing and data analysis workflows in Python.

```python
import numpy as np

# Create a NumPy array
arr = np.array([1, 2, 3, 4, 5])

# Perform a mathematical operation (example: square each element)
squared_arr = np.square(arr)

print(squared_arr)



```
In this example:

np.array([1, 2, 3, 4, 5]) creates a NumPy array.
np.square(arr) computes the square of each element in the array using NumPy's vectorized operations.

In [2]:
arr= np.array([25,41,63,66,85,74])
arr

array([25, 41, 63, 66, 85, 74])

In [3]:
type(arr)
#ndarray ==>n dimensiional array

numpy.ndarray

In [4]:
arr.size #no.of item

6

In [5]:
len(arr)

6

In [6]:
arr.ndim # to check the dimension of array

1

In [7]:
arr= np.array([25,41,63,66,85,74,25.01,"upflairs",True])
arr

array(['25', '41', '63', '66', '85', '74', '25.01', 'upflairs', 'True'],
      dtype='<U32')

In [8]:
arr.dtype


dtype('<U32')

In [9]:
np.__version__


'2.0.0'

The expression np.__version__ is used to retrieve the version number of the NumPy library that is currently imported into your Python environment. Here’s what it does:

np: This refers to the alias np that you assigned to the NumPy library when you imported it using import numpy as np.
.__version__: This is a special attribute of the NumPy module that holds the version information of the library.

In [10]:
arr= np.array([25,41,63,66,85,74])
arr

array([25, 41, 63, 66, 85, 74])

In [11]:
# arr[-1]=500
arr[2:4]
arr[::]# start=0,stoping =end,jump==1

array([25, 41, 63, 66, 85, 74])

In [12]:
arr*2

array([ 50,  82, 126, 132, 170, 148])

In [13]:
arr[-2]


np.int64(85)

In [14]:
arr[::]# start=0,stoping =end,jump==1


array([25, 41, 63, 66, 85, 74])

# 2D  ARRAYs


In [15]:
ls=[[1,2,3],
    [4,5,6],
    [7,8,9]]

In [16]:
arr2 =np.array(ls)
print(arr2)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [17]:
type(ls)

list

In [18]:
arr2.ndim

2

In [19]:
# arr2[0]# for single row only
arr2[0:2]#start to end for more than one row

array([[1, 2, 3],
       [4, 5, 6]])

In [20]:

# arr2[row,column]=>[row start:stop,col start:end]
arr2[1,1:]

array([5, 6])

In [21]:
# arr2[1:3,1:]
arr2[-2:,-2:]

array([[5, 6],
       [8, 9]])

In [22]:
arr2[-1:-1]=90

### filter all those items that are less than equals to 100
### comments --> how many item in your arrays

In [23]:
arr= np.random.randint(1,200,500)
arr[arr<=100]

array([ 26,  76,  85,  34,  15,   1,  58,  39,  62,  55,  31,  18,  84,
        30,  10,  60,   9,  91,  35,  99,  61,  45,  47,  98,  98,  65,
        48,  92,  24,  49,  99,  25,  72,  58,  75,  75,  20,  32,  77,
         1,  76,  43,  94,   1,  47,  11,   8,  23,  69,  75,  84,  62,
        25,  33,  40,  26,  42,  87, 100,  96,  60,  90,  56,  28,  48,
        48,  81,  10,  48,  55,  33,   1,  57,  65,   6,  76,  69,  44,
        28,  51,  79,  98,  10,   5,  71,  49,  70,  21,  97,  39,  23,
        16,  64,  29,  60,  41,  42,  90,  93,  22,  92,  16,  21,  93,
         9,  62,  53,   6,  78,  12,  91, 100,  20,  62,  10,  69,  70,
        59,  19,  95,  30,  52,  17,  38,  19,  26,  84,  57,  41,  54,
        31,  78,  56,  16,  27,  27,  35,  94,  33,  83,  26,  51,  27,
        24,  75,  95,  35,  14,  50,  95,  27,  75,  68,  24,  19,  88,
        68,  62,  38,  14,  42,  35,   4,  12,  79,  71,  97,  13,  62,
        39,  72,  34,  68,  27,   1,  62,  76,  51,   6,  71,   

In [24]:
# for item in arr:
#     count=0
#     if item<=100:
#         count+=1
# print(count)        

In [25]:
len(arr[arr<=100])

247

In [26]:
# arr.shape(3,3)

In [27]:
arr2=np.random.randint(1,200,(10,5))#range 1 to 200 
# and the rows =10 and columns =5
arr2

array([[138, 132,  57, 162,  93],
       [  4,  55, 116,  17, 161],
       [ 98,   1,  23, 160,  23],
       [110, 122, 184,  65,  25],
       [ 13,  67, 132, 150,  55],
       [102, 183, 137, 176,  54],
       [165, 161,  51,  38,  32],
       [176,  34,  77,  36,  79],
       [ 58,   6, 143,  36, 172],
       [ 74,  56,  66,  56,  73]], dtype=int32)

In [28]:
arr=np.zeros((10,5))
arr

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [29]:
arr.dtype

dtype('float64')

In [30]:
arr =np.ones((10,5))
arr

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

In [31]:
ls=[0,1,2,3,4,5,6,7,8,9]
ls

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [32]:
ls=list(range(10))
ls

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [33]:
arr=np.arange(20)
arr# always create single dimension array in sorted manner from 0 to 20/range we have given

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19])

In [34]:
arr=np.arange(60)
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59])

In [35]:
arr.reshape(10,6)#10(rows)*6(columns)==60
# arr.reshape(10,7)#this will show error because 10*7==70 which is invalis=d as it exceed the range of it

array([[ 0,  1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10, 11],
       [12, 13, 14, 15, 16, 17],
       [18, 19, 20, 21, 22, 23],
       [24, 25, 26, 27, 28, 29],
       [30, 31, 32, 33, 34, 35],
       [36, 37, 38, 39, 40, 41],
       [42, 43, 44, 45, 46, 47],
       [48, 49, 50, 51, 52, 53],
       [54, 55, 56, 57, 58, 59]])

In [36]:
arr.reshape(30,2)

array([[ 0,  1],
       [ 2,  3],
       [ 4,  5],
       [ 6,  7],
       [ 8,  9],
       [10, 11],
       [12, 13],
       [14, 15],
       [16, 17],
       [18, 19],
       [20, 21],
       [22, 23],
       [24, 25],
       [26, 27],
       [28, 29],
       [30, 31],
       [32, 33],
       [34, 35],
       [36, 37],
       [38, 39],
       [40, 41],
       [42, 43],
       [44, 45],
       [46, 47],
       [48, 49],
       [50, 51],
       [52, 53],
       [54, 55],
       [56, 57],
       [58, 59]])


The reshape method in NumPy is used to change the shape (dimensions) of an array without changing its data. Here’s a detailed explanation of how arr.reshape works and its usage:

Syntax and Parameters
The reshape method in NumPy typically takes one or more arguments:

```python

arr.reshape(shape, order='C')
```
arr: This is the NumPy array that you want to reshape.
shape: This parameter specifies the new shape that you want the array to have. It can be specified in several formats:
A single integer, which means the resulting array will have that many dimensions.
A tuple of integers, where each integer specifies the size of the corresponding dimension in the reshaped array. The total number of elements in the reshaped array must match the original array.
order: This is an optional parameter that specifies the order of the elements in the reshaped array. It can be 'C' (row-major, C-style) or 'F' (column-major, Fortran-style).
How reshape Works
Shape Compatibility: The total number of elements in the original array (arr) must be equal to the total number of elements in the reshaped array specified by shape. NumPy will raise an error if these numbers do not match.

Data View: reshape does not copy the data; it returns a new view of the original array with the specified shape. This means that modifying the reshaped array will also modify the original array.

# CREATION OF 3D ARRAY


In [37]:
arr1=[1,2,3] # 1d array
arr2=[[1,2,3],[1,2,3],[1,2,3]] #2d array
arr3 = [
    [[1, 2, 3], [1, 2, 3], [1, 2, 3]],
    [[1, 2, 3], [1, 2, 3], [1, 2, 3]]
]#3d array


In [38]:
import numpy as np

In [39]:
arr3=np.random.randint(1,200,(3,5,4))
arr3

array([[[ 13,  86, 148,  49],
        [191, 111, 135, 172],
        [ 71, 172,  91, 199],
        [ 78, 125,  60,  98],
        [ 79, 117, 150,  12]],

       [[105, 104, 166, 137],
        [128, 176,  49, 118],
        [177, 195, 162,  20],
        [ 99,  37,  82, 160],
        [ 89,   2, 107, 153]],

       [[192, 163, 144,  44],
        [159,   7, 184,  25],
        [123, 157,  38, 135],
        [ 36, 133, 170, 113],
        [ 36, 168, 103,  11]]], dtype=int32)

The code `arr3 = np.random.randint(1, 200, (3, 5, 4))` uses NumPy to create a 3-dimensional array named `arr3`. This array has random integers between 1 and 199 (inclusive). The shape of the array is `(3, 5, 4)`, meaning it has:

- 3 layers (first dimension)
- 5 rows in each layer (second dimension)
- 4 columns in each row (third dimension)

Here's an example visualization of the structure:

```
Layer 1:
[[a, b, c, d],
 [e, f, g, h],
 [i, j, k, l],
 [m, n, o, p],
 [q, r, s, t]]

Layer 2:
[[u, v, w, x],
 [y, z, aa, bb],
 [cc, dd, ee, ff],
 [gg, hh, ii, jj],
 [kk, ll, mm, nn]]

Layer 3:
[[oo, pp, qq, rr],
 [ss, tt, uu, vv],
 [ww, xx, yy, zz],
 [aaa, bbb, ccc, ddd],
 [eee, fff, ggg, hhh]]
```

Each letter represents a random integer.

In [40]:
arr3.ndim

3

In [41]:
arr3.dtype

dtype('int32')

In [42]:
arr3.shape

(3, 5, 4)

### how to access the value from the array 
*arr[table,row,column]<- indexing method

In [43]:
arr3[1,4,0:]

array([ 89,   2, 107, 153], dtype=int32)

In [44]:
arr3[2,0:2,1:3]

array([[163, 144],
       [  7, 184]], dtype=int32)

In [45]:
# arr3[0,0:1,0:]
arr3[-2,-5:-4,-5:]

array([[105, 104, 166, 137]], dtype=int32)

In [46]:
arr3[-1,-5:-4,-5:]

array([[192, 163, 144,  44]], dtype=int32)

# OTHER FUNCTIONS IN NUMPY AND THEIR FUNCTIONALITY

In [47]:
arr=np.array([9,7,8])
arr

array([9, 7, 8])

In [48]:
max(arr)

np.int64(9)

In [49]:
min(arr)

np.int64(7)

In [50]:
np.mean(arr)# average of the array

np.float64(8.0)

In [51]:
np.sum(arr3)

np.int64(6564)

In [52]:
np.argmin(arr3)# this show the the index of the minimum number


np.int64(37)

In [53]:
np.argmax(arr)# this shows the index of the maximum item in array 

np.int64(0)

In [54]:
arr.sort()
arr

array([7, 8, 9])

In [55]:
arr[::-1]#descending order

array([9, 8, 7])

# try with 4D  and 5-D array homework

In [56]:
import numpy as np

# Create a 4D array with shape (2, 3, 4, 5)
# Here, 2 is the number of 3D blocks, each 3D block has shape (4, 5)
array4 = np.random.randint(1, 10, (2, 3, 4, 5))

print(array4)


[[[[7 8 4 4 5]
   [2 6 2 5 8]
   [6 3 3 4 8]
   [8 5 7 3 8]]

  [[9 9 3 9 8]
   [7 8 6 2 8]
   [2 1 4 2 3]
   [4 9 2 7 9]]

  [[2 9 8 2 2]
   [7 1 5 9 4]
   [7 2 1 4 4]
   [9 6 2 3 3]]]


 [[[3 9 9 6 8]
   [1 5 6 5 3]
   [2 8 7 8 6]
   [2 1 3 1 7]]

  [[7 8 3 6 4]
   [1 1 2 9 8]
   [6 6 9 6 1]
   [8 2 8 9 1]]

  [[7 6 8 4 7]
   [8 8 4 1 6]
   [1 1 5 9 8]
   [7 8 4 6 3]]]]


In [57]:
array4[1,2,3,1:2]

array([8], dtype=int32)